cudax/include/cuda/experimental/__async/basic_sender.cuh

```diff
@@ -0,0 +1,242 @@
+
+  _CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE decltype(auto) get_env() const noexcept
+  {
+    return _rcvr_t::get_env(state_.data, state_.receiver);
+  }
+};
+
+template <class Rcvr>
+_CCCL_INLINE_VAR constexpr bool has_no_environment = _CUDA_VSTD::is_same_v<Rcvr, receiver_archetype>;
+
+template <bool HasStopped, class Data, class Rcvr>
+struct _mk_completions
+{
+  using _rcvr_t = typename Data::receiver_tag;
+
+  template <class... Args>
+  using _set_value_t =
+    decltype(+*_rcvr_t::set_value(_declval<Data&>(), _declval<receiver_archetype&>(), _declval<Args>()...));
+
+  template <class Error>
+  using _set_error_t =
+    decltype(+*_rcvr_t::set_error(_declval<Data&>(), _declval<receiver_archetype&>(), _declval<Error>()));
+
+  using _set_stopped_t = __async::completion_signatures<>;
+};
+
+template <class Data, class Rcvr>
+struct _mk_completions<true, Data, Rcvr> : _mk_completions<false, Data, Rcvr>
+{
+  using _rcvr_t = typename Data::receiver_tag;
+
+  using _set_stopped_t = decltype(+*_rcvr_t::set_stopped(_declval<Data&>(), _declval<receiver_archetype&>()));
+};
+
+template <class...>
+using _ignore_value_signature = __async::completion_signatures<>;
+
+template <class>
+using _ignore_error_signature = __async::completion_signatures<>;
+
+template <class Completions>
+constexpr bool _has_stopped =
+  !_CUDA_VSTD::is_same_v<__async::completion_signatures<>,
+                         __async::transform_completion_signatures<Completions,
+                                                                  __async::completion_signatures<>,
+                                                                  _ignore_value_signature,
+                                                                  _ignore_error_signature>>;
+
+template <bool PotentiallyThrowing, class Rcvr>
+void set_current_exception_if(Rcvr& rcvr) noexcept
+{
+  if constexpr (PotentiallyThrowing)
+  {
+    __async::set_error(_CUDA_VSTD::move(rcvr), ::std::current_exception());
+  }
+}
+
+// A generic type that holds the data for an async operation, and
+// that provides a `start` method for enqueuing the work.
+template <class Sndr, class Data, class Rcvr>
+struct basic_opstate
+{
+  using _rcvr_t        = basic_receiver<Data, Rcvr>;
+  using _completions_t = completion_signatures_of_t<Sndr, _rcvr_t>;
+  using _traits_t      = _mk_completions<_has_stopped<_completions_t>, Data, Rcvr>;
+
+  using completion_signatures =
+    transform_completion_signatures<_completions_t,
+                                    __async::completion_signatures<>, // TODO
+                                    _traits_t::template _set_value_t,
+                                    _traits_t::template _set_error_t,
+                                    typename _traits_t::_set_stopped_t>;
+
+  _CCCL_HOST_DEVICE basic_opstate(Sndr&& sndr, Data data, Rcvr rcvr)
+      : state_{static_cast<Data&&>(data), static_cast<Rcvr&&>(rcvr)}
+      , op_(__async::connect(static_cast<Sndr&&>(sndr), _rcvr_t{state_}))
```

Comment:
This uses `static_cast` instead of `_CUDA_VSTD::move`, I believe we should settle on the latter

Reply:
i thought this would come up. this is a pretty template-heavy library. i go to lengths to instantiate as few templates as possible. not all compilers treat `std::move` and `std::forward` as intrinsics. they are just casts, so instantiating a function template for this seems wasteful to me. thoughts?

Reply:
I am all in favor of better build times, but then we should be consistent

Reply:
i assume you mean: use `move` and `forward` everywhere, or use `static_cast` everywhere, right?

where is "everywhere" tho? everywhere within the `cudax/experimental/__async` directory? or all of cudax? or all of cccl?

Reply:
I would say anywhere within the `cudax/experimental/__async`